REGRESSION ( Sonoma?): [ Sonoma release x86_64 ] Three TestWebKitAPI.WKWebExtensionAPITabs tests are a consistent timeout

xeenon opened the pull request:
#### 8cd58dac3ab9b34d9c0cf79fd53770fed42614a8
<pre>
REGRESSION ( Sonoma?): [ Sonoma release x86_64 ] Three TestWebKitAPI.WKWebExtensionAPITabs tests are a consistent timeout
<a href="https://webkit.org/b/263525">https://webkit.org/b/263525</a>
<a href="https://rdar.apple.com/problem/117354451">rdar://problem/117354451</a>

Reviewed by Alex Christensen.

Add EagerCallbackAggregator, which is like CallbackAggregator, but ensures a callback is executed at
its first opportunity, ignoring subsequent triggers. This was needed for the reply mechanism used
by runtime.sendMessage and friends.

* Source/WTF/wtf/CallbackAggregator.h:
(WTF::EagerCallbackAggregator): Added.
* Source/WTF/wtf/Ref.h:
(WTF::Ref::operator* const): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIRuntimeCocoa.mm:
(WebKit::WebExtensionContext::runtimeSendMessage): Use EagerCallbackAggregator.

Canonical link: <a href="https://commits.webkit.org/271636@main">https://commits.webkit.org/271636@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/92c60937f0ea28c9c8a1166e06fe348a9531bdc0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29118 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7786 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30455 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/31717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/26500 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/9912 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5092 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/31717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29390 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/9912 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/24953 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/31717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/9912 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/25986 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/33056 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/25121 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/9912 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/26420 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/31947 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/29441 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/5686 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/3853 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/29731 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/7351 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/35728 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/6190 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/7697 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3737 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/6207 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WTF/wtf/CallbackAggregator.h:
```diff
@@ -62,7 +63,55 @@ class CallbackAggregatorOnThread : public ThreadSafeRefCounted<CallbackAggregato
+    Out operator()(In... args)
+    {
+        if (m_callback) {
+            if constexpr (std::is_same<Out, void>::value) {
```
I don't think this is needed.  You can return the result of the callback if Out is void.

achristensen07 commented on Source/WTF/wtf/CallbackAggregator.h:
```diff
@@ -62,7 +63,55 @@ class CallbackAggregatorOnThread : public ThreadSafeRefCounted<CallbackAggregato
+class EagerCallbackAggregator<Out(In...)> : public RefCounted<EagerCallbackAggregator<Out(In...)>> {
+public:
+    template<typename CallableType, class = typename std::enable_if<std::is_rvalue_reference<CallableType&&>::value>::type>
+    static RefPtr<EagerCallbackAggregator> create(CallableType&& callback, In... defaultArgs)
```
This can return a Ref.

achristensen07 commented on Source/WTF/wtf/CallbackAggregator.h:
```diff
@@ -62,7 +63,55 @@ class CallbackAggregatorOnThread : public ThreadSafeRefCounted<CallbackAggregato
+    {
+    }
+
+    Function<Out(In...)> m_callback;
```
This could be a CompletionHandler.  Then you wouldn't need to do all the std::exchange(m_callback, nullptr) stuff above.

achristensen07 commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIRuntimeCocoa.mm:
```diff
@@ -133,18 +136,13 @@
-    for (auto& process : mainWorldProcesses)
-        process->sendWithAsyncReply(Messages::WebExtensionContextProxy::DispatchRuntimeMessageEvent(WebExtensionContentWorldType::Main, messageJSON, std::nullopt, completeSenderParameters), handleReply, identifier());
+    for (auto& process : mainWorldProcesses) {
+        process->sendWithAsyncReply(Messages::WebExtensionContextProxy::DispatchRuntimeMessageEvent(WebExtensionContentWorldType::Main, messageJSON, std::nullopt, completeSenderParameters), [callbackAggregator, protectedThis = Ref { *this }](std::optional<String> replyJSON) {
```
protectedThis is likely not necessary.  It's not used.

xeenon committed:
REGRESSION ( Sonoma?): [ Sonoma release x86_64 ] Three TestWebKitAPI.WKWebExtensionAPITabs tests are a consistent timeout
https://webkit.org/b/263525
rdar://problem/117354451

Reviewed by Alex Christensen.

Add EagerCallbackAggregator, which is like CallbackAggregator, but ensures a callback is executed at
its first opportunity, ignoring subsequent triggers. This was needed for the reply mechanism used
by runtime.sendMessage and friends.

* Source/WTF/wtf/CallbackAggregator.h:
(WTF::EagerCallbackAggregator): Added.
* Source/WTF/wtf/Ref.h:
(WTF::Ref::operator* const): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIRuntimeCocoa.mm:
(WebKit::WebExtensionContext::runtimeSendMessage): Use EagerCallbackAggregator.

Canonical link: https://commits.webkit.org/271636@main
